soifunc/resize.py

```diff
@@ -1,70 +1,42 @@
-from typing import List, Optional, Union
-
-import muvsfunc
-import vapoursynth as vs
-import vsutil
-from nnedi3_resample import nnedi3_resample
-
-from .internal import value_error
-
-core = vs.core
-
-__all__ = ["GoodResize", "Descale", "DescaleMask"]
-
-
-def GoodResize(
+import warnings
+from dataclasses import dataclass
+from inspect import getfullargspec
+from typing import Any, cast
+
+from vskernels import Kernel, Scaler, ScalerT, Spline36
+from vsscale import SSIM, GenericScaler
+from vsscale import descale as descale_iew
+from vsscale import descale_detail_mask
+from vstools import check_variable_format, join, vs
+
+__all__ = [
+    "good_resize", "GoodResize",
+    "descale", "Descale",
+    "descale_mask", "DescaleMask",
+    "HybridScaler"
+]
+
+
+def good_resize(
     clip: vs.VideoNode,
     width: int,
     height: int,
     gpu: bool = False,
-    device: Optional[int] = None,
 ) -> vs.VideoNode:
     """High quality resizing filter"""
-    if clip.width == width and clip.height == height:
+    from vsaa import Nnedi3
+
+    if (width, height) == (clip.width, clip.height):
         return clip
-    planes: List[vs.VideoNode] = vsutil.split(clip)
-    upscale = width >= clip.width or height >= clip.height
-
-    for i in range(len(planes)):
-        if i == 0:
-            if upscale:
-                planes[0] = nnedi3_resample(
-                    planes[0],
-                    width,
-                    height,
-                    mode="nnedi3cl" if gpu else "znedi3",
-                    nsize=4,
-                    nns=4,
-                    device=device,
-                )
-            else:
-                planes[0] = muvsfunc.SSIM_downsample(
-                    planes[0],
-                    width,
-                    height,
-                    kernel="Lanczos",
-                    smooth=0.5,
-                    dither_type="error_diffusion",
-                )
-                planes[0] = vsutil.depth(planes[0], clip.format.bits_per_sample)
-        else:
-            planes[i] = planes[i].resize.Spline36(
-                width >> clip.format.subsampling_w,
-                height >> clip.format.subsampling_h,
-                dither_type="error_diffusion",
-            )
-
-    if len(planes) == 1:
-        return planes[0]
-
-    return vsutil.join(planes, clip.format.color_family)
-
-
-def Descale(
+
+    return Nnedi3(opencl=gpu, scaler=HybridScaler(SSIM, Spline36)).scale(clip, width, height)
```

Comment:
From the code it looks like if we are purely downscaling, then Nnedi3 will not do any upscaling and just straight downscale with the HybridScaler, can you confirm that is correct?